File: starter-code/session-6-go-sample/internal/service/fraud_detection_service.go
package service

import (
	"fraud-detection-service/internal/models"
	"math"
	"strings"
	"sync"
	"time"
)

// FraudDetectionService implements fraud detection business logic
type FraudDetectionService struct {
	// In-memory storage for demo purposes (would be replaced with proper database)
	accountHistory      map[string][]models.Transaction
	dailyTransactionCounts map[string]int
	mutex               sync.RWMutex
}

// Configuration constants
const (
	HighAmountThreshold       = 10000.0
	UnusualAmountMultiplier   = 5.0
	MaxDailyTransactions      = 10
	SuspiciousVelocityMinutes = 5
)

// NewFraudDetectionService creates a new fraud detection service
func NewFraudDetectionService() *FraudDetectionService {
	return &FraudDetectionService{
		accountHistory:         make(map[string][]models.Transaction),
		dailyTransactionCounts: make(map[string]int),
	}
}

// AnalyzeTransaction performs fraud analysis on a transaction
func (s *FraudDetectionService) AnalyzeTransaction(transaction models.Transaction) (*models.FraudAnalysisResult, error) {
	// Validate transaction
	if err := transaction.Validate(); err != nil {
		return nil, err
	}

	var riskScore float64
	var riskFactors []string
	var recommendations []string

	// Amount-based risk analysis
	riskScore += s.analyzeAmountRisk(transaction, &riskFactors)

	// Velocity-based risk analysis
	riskScore += s.analyzeVelocityRisk(transaction, &riskFactors)

	// Geographic risk analysis
	riskScore += s.analyzeGeographicRisk(transaction, &riskFactors)

	// Time-based risk analysis
	riskScore += s.analyzeTimeBasedRisk(transaction, &riskFactors)

	// Device and IP analysis
	riskScore += s.analyzeDeviceRisk(transaction, &riskFactors)

	// Merchant category analysis
	riskScore += s.analyzeMerchantRisk(transaction, &riskFactors)

	// Normalize risk score to 0-1 range
	riskScore = math.Min(1.0, riskScore)

	// Determine risk level and decision
	riskLevel := models.FromScore(riskScore)
	decision := s.determineDecision(riskScore)

	// Generate recommendations
	s.generateRecommendations(riskLevel, &recommendations)

	// Store transaction in history for future analysis
	s.updateTransactionHistory(transaction)

	// Create and return result
	result := models.NewFraudAnalysisResult(
		transaction.TransactionID,
		riskScore,
		riskLevel,
		decision,
	)
	result.RiskFactors = riskFactors
	result.Recommendations = recommendations

	return result, nil
}

// analyzeAmountRisk performs amount-based risk analysis
func (s *FraudDetectionService) analyzeAmountRisk(transaction models.Transaction, riskFactors *[]string) float64 {
	risk := 0.0

	// High amount transactions are inherently riskier
	if transaction.Amount > HighAmountThreshold {
		risk += 0.3
		*riskFactors = append(*riskFactors, "High transaction amount")
	}

	// Check if amount is unusual for this account
	s.mutex.RLock()
	history, exists := s.accountHistory[transaction.AccountID]
	s.mutex.RUnlock()

	if exists && len(history) > 0 {
		var sum float64
		for _, t := range history {
			sum += t.Amount
		}
		avgAmount := sum / float64(len(history))

		if transaction.Amount > avgAmount*UnusualAmountMultiplier {
			risk += 0.2
			*riskFactors = append(*riskFactors, "Unusual amount compared to account history")
		}
	}

	return risk
}

// analyzeVelocityRisk performs velocity-based risk analysis
func (s *FraudDetectionService) analyzeVelocityRisk(transaction models.Transaction, riskFactors *[]string) float64 {
	risk := 0.0

	// Check daily transaction count
	dailyKey := transaction.AccountID + ":" + transaction.TransactionTimestamp.Format("2006-01-02")
	
	s.mutex.Lock()
	dailyCount := s.dailyTransactionCounts[dailyKey] + 1
	s.dailyTransactionCounts[dailyKey] = dailyCount
	s.mutex.Unlock()

	if dailyCount > MaxDailyTransactions {
		risk += 0.2
		*riskFactors = append(*riskFactors, "High daily transaction count")
	}

	// Check for rapid successive transactions
	recentTransactions := s.getRecentTransactions(transaction.AccountID, SuspiciousVelocityMinutes)
	if len(recentTransactions) >= 3 {
		risk += 0.25
		*riskFactors = append(*riskFactors, "Multiple transactions within short time frame")
	}

	return risk
}

// analyzeGeographicRisk performs geographic risk analysis
func (s *FraudDetectionService) analyzeGeographicRisk(transaction models.Transaction, riskFactors *[]string) float64 {
	risk := 0.0

	// Check for high-risk countries
	if s.isHighRiskCountry(transaction.LocationCountry) {
		risk += 0.4
		*riskFactors = append(*riskFactors, "Transaction from high-risk country: "+transaction.LocationCountry)
	}

	// Check for geographic inconsistency (simplified implementation)
	if s.hasGeographicInconsistency(transaction) {
		risk += 0.3
		*riskFactors = append(*riskFactors, "Geographic inconsistency detected")
	}

	return risk
}

// analyzeTimeBasedRisk performs time-based risk analysis
func (s *FraudDetectionService) analyzeTimeBasedRisk(transaction models.Transaction, riskFactors *[]string) float64 {
	risk := 0.0

	// Transactions during unusual hours
	hour := transaction.TransactionTimestamp.Hour()
	if hour < 6 || hour > 23 {
		risk += 0.1
		*riskFactors = append(*riskFactors, "Transaction during unusual hours")
	}

	return risk
}

// analyzeDeviceRisk performs device and IP-based risk analysis
func (s *FraudDetectionService) analyzeDeviceRisk(transaction models.Transaction, riskFactors *[]string) float64 {
	risk := 0.0

	// Check for new device
	if transaction.DeviceFingerprint != "" && s.isNewDevice(transaction) {
		risk += 0.15
		*riskFactors = append(*riskFactors, "Transaction from new device")
	}

	// Check for suspicious IP
	if s.isSuspiciousIP(transaction.IPAddress) {
		risk += 0.2
		*riskFactors = append(*riskFactors, "Transaction from suspicious IP address")
	}

	return risk
}

// analyzeMerchantRisk performs merchant-related risk analysis
func (s *FraudDetectionService) analyzeMerchantRisk(transaction models.Transaction, riskFactors *[]string) float64 {
	risk := 0.0

	// High-risk merchant categories
	if s.isHighRiskMerchantCategory(transaction.MerchantCategory) {
		risk += 0.15
		*riskFactors = append(*riskFactors, "High-risk merchant category: "+transaction.MerchantCategory)
	}

	return risk
}

// determineDecision determines the final decision based on risk score
func (s *FraudDetectionService) determineDecision(riskScore float64) string {
	if riskScore < 0.3 {
		return "APPROVE"
	} else if riskScore < 0.7 {
		return "REVIEW"
	}
	return "DECLINE"
}

// generateRecommendations generates recommendations based on risk level
func (s *FraudDetectionService) generateRecommendations(riskLevel models.RiskLevel, recommendations *[]string) {
	switch riskLevel {
	case models.LowRisk:
		*recommendations = append(*recommendations, "Process transaction normally")
	case models.MediumRisk:
		*recommendations = append(*recommendations, "Additional verification recommended")
		*recommendations = append(*recommendations, "Monitor account for suspicious activity")
	case models.HighRisk:
		*recommendations = append(*recommendations, "Manual review required")
		*recommendations = append(*recommendations, "Contact customer for verification")
		*recommendations = append(*recommendations, "Consider temporary account restrictions")
	}
}

// Helper methods

func (s *FraudDetectionService) updateTransactionHistory(transaction models.Transaction) {
	s.mutex.Lock()
	defer s.mutex.Unlock()
	
	s.accountHistory[transaction.AccountID] = append(
		s.accountHistory[transaction.AccountID], 
		transaction,
	)
}

func (s *FraudDetectionService) getRecentTransactions(accountID string, minutes int) []models.Transaction {
	s.mutex.RLock()
	defer s.mutex.RUnlock()
	
	transactions, exists := s.accountHistory[accountID]
	if !exists {
		return []models.Transaction{}
	}

	cutoff := time.Now().Add(-time.Duration(minutes) * time.Minute)
	var recent []models.Transaction
	
	for _, t := range transactions {
		if t.TransactionTimestamp.After(cutoff) {
			recent = append(recent, t)
		}
	}
	
	return recent
}

func (s *FraudDetectionService) isHighRiskCountry(country string) bool {
	// Simplified implementation
	highRiskCountries := []string{"XX", "YY", "ZZ"} // Placeholder countries
	for _, hrc := range highRiskCountries {
		if country == hrc {
			return true
		}
	}
	return false
}

func (s *FraudDetectionService) hasGeographicInconsistency(transaction models.Transaction) bool {
	// Simplified implementation - would check against recent transactions
	return false // Placeholder implementation
}

func (s *FraudDetectionService) isNewDevice(transaction models.Transaction) bool {
	// Simplified implementation - would check against known devices
	return true // Placeholder - always return true for demo
}

func (s *FraudDetectionService) isSuspiciousIP(ipAddress string) bool {
	// Simplified implementation
	return strings.HasPrefix(ipAddress, "10.0.0") // Placeholder logic
}

func (s *FraudDetectionService) isHighRiskMerchantCategory(category string) bool {
	// High-risk categories
	highRiskCategories := []string{"gambling", "adult", "cryptocurrency", "money_transfer"}
	for _, hrc := range highRiskCategories {
		if category == hrc {
			return true
		}
	}
	return false
}

File: starter-code/session-6-java-module/src/main/java/com/megabank/fraud/service/FraudDetectionService.java
package com.megabank.fraud.service;

import com.megabank.fraud.model.*;
import org.springframework.stereotype.Service;
import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.time.temporal.ChronoUnit;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Core fraud detection service implementing risk analysis algorithms
 * This service contains the main business logic for fraud detection
 */
@Service
public class FraudDetectionService {
    
    // In-memory cache for transaction history (in real system, this would be a database)
    private final Map<String, List<Transaction>> accountTransactionHistory = new ConcurrentHashMap<>();
    private final Map<String, Integer> dailyTransactionCounts = new ConcurrentHashMap<>();
    
    private static final double HIGH_AMOUNT_THRESHOLD = 10000.0;
    private static final double UNUSUAL_AMOUNT_MULTIPLIER = 5.0;
    private static final int MAX_DAILY_TRANSACTIONS = 10;
    private static final int SUSPICIOUS_VELOCITY_MINUTES = 5;
    
    /**
     * Main method to analyze a transaction for fraud
     */
    public FraudAnalysisResult analyzeTransaction(Transaction transaction) {
        double riskScore = 0.0;
        List<String> riskFactors = new ArrayList<>();
        List<String> recommendations = new ArrayList<>();
        
        // Amount-based risk analysis
        riskScore += analyzeAmountRisk(transaction, riskFactors);
        
        // Velocity-based risk analysis
        riskScore += analyzeVelocityRisk(transaction, riskFactors);
        
        // Geographic risk analysis
        riskScore += analyzeGeographicRisk(transaction, riskFactors);
        
        // Time-based risk analysis
        riskScore += analyzeTimeBasedRisk(transaction, riskFactors);
        
        // Device and IP analysis
        riskScore += analyzeDeviceRisk(transaction, riskFactors);
        
        // Merchant category analysis
        riskScore += analyzeMerchantRisk(transaction, riskFactors);
        
        // Normalize risk score to 0-1 range
        riskScore = Math.min(1.0, riskScore);
        
        // Determine risk level and decision
        RiskLevel riskLevel = RiskLevel.fromScore(riskScore);
        String decision = determineDecision(riskScore);
        
        // Generate recommendations
        generateRecommendations(riskScore, riskLevel, recommendations);
        
        // Store transaction in history for future analysis
        updateTransactionHistory(transaction);
        
        FraudAnalysisResult result = new FraudAnalysisResult(
            transaction.getTransactionId(), 
            riskScore, 
            riskLevel, 
            decision
        );
        result.setRiskFactors(riskFactors);
        result.setRecommendations(recommendations);
        result.setModelVersion("1.0");
        
        return result;
    }
    
    /**
     * Analyze risk based on transaction amount
     */
    private double analyzeAmountRisk(Transaction transaction, List<String> riskFactors) {
        double risk = 0.0;
        BigDecimal amount = transaction.getAmount();
        
        // High amount transactions are inherently riskier
        if (amount.compareTo(BigDecimal.valueOf(HIGH_AMOUNT_THRESHOLD)) > 0) {
            risk += 0.3;
            riskFactors.add("High transaction amount: $" + amount);
        }
        
        // Check if amount is unusual for this account
        List<Transaction> history = accountTransactionHistory.get(transaction.getAccountId());
        if (history != null && !history.isEmpty()) {
            double avgAmount = history.stream()
                .mapToDouble(t -> t.getAmount().doubleValue())
                .average()
                .orElse(0.0);
            
            if (amount.doubleValue() > avgAmount * UNUSUAL_AMOUNT_MULTIPLIER) {
                risk += 0.2;
                riskFactors.add("Unusual amount compared to account history");
            }
        }
        
        return risk;
    }
    
    /**
     * Analyze risk based on transaction velocity
     */
    private double analyzeVelocityRisk(Transaction transaction, List<String> riskFactors) {
        double risk = 0.0;
        String accountId = transaction.getAccountId();
        
        // Check daily transaction count
        String dailyKey = accountId + ":" + transaction.getTransactionTimestamp().toLocalDate();
        int dailyCount = dailyTransactionCounts.getOrDefault(dailyKey, 0) + 1;
        dailyTransactionCounts.put(dailyKey, dailyCount);
        
        if (dailyCount > MAX_DAILY_TRANSACTIONS) {
            risk += 0.2;
            riskFactors.add("High daily transaction count: " + dailyCount);
        }
        
        // Check for rapid successive transactions
        List<Transaction> recentTransactions = getRecentTransactions(accountId, SUSPICIOUS_VELOCITY_MINUTES);
        if (recentTransactions.size() >= 3) {
            risk += 0.25;
            riskFactors.add("Multiple transactions within " + SUSPICIOUS_VELOCITY_MINUTES + " minutes");
        }
        
        return risk;
    }
    
    /**
     * Analyze geographic risk factors
     */
    private double analyzeGeographicRisk(Transaction transaction, List<String> riskFactors) {
        double risk = 0.0;
        
        // Check for high-risk countries
        if (isHighRiskCountry(transaction.getLocationCountry())) {
            risk += 0.4;
            riskFactors.add("Transaction from high-risk country: " + transaction.getLocationCountry());
        }
        
        // Check for geographic inconsistency with recent transactions
        if (hasGeographicInconsistency(transaction)) {
            risk += 0.3;
            riskFactors.add("Geographic inconsistency detected");
        }
        
        return risk;
    }
    
    /**
     * Analyze time-based risk factors
     */
    private double analyzeTimeBasedRisk(Transaction transaction, List<String> riskFactors) {
        double risk = 0.0;
        LocalDateTime timestamp = transaction.getTransactionTimestamp();
        
        // Transactions during unusual hours
        int hour = timestamp.getHour();
        if (hour < 6 || hour > 23) {
            risk += 0.1;
            riskFactors.add("Transaction during unusual hours");
        }
        
        return risk;
    }
    
    /**
     * Analyze device and IP-based risk
     */
    private double analyzeDeviceRisk(Transaction transaction, List<String> riskFactors) {
        double risk = 0.0;
        
        // Check for new device
        if (transaction.getDeviceFingerprint() != null && isNewDevice(transaction)) {
            risk += 0.15;
            riskFactors.add("Transaction from new device");
        }
        
        // Check for suspicious IP
        if (isSuspiciousIP(transaction.getIpAddress())) {
            risk += 0.2;
            riskFactors.add("Transaction from suspicious IP address");
        }
        
        return risk;
    }
    
    /**
     * Analyze merchant-related risk
     */
    private double analyzeMerchantRisk(Transaction transaction, List<String> riskFactors) {
        double risk = 0.0;
        
        // High-risk merchant categories
        if (isHighRiskMerchantCategory(transaction.getMerchantCategory())) {
            risk += 0.15;
            riskFactors.add("High-risk merchant category: " + transaction.getMerchantCategory());
        }
        
        return risk;
    }
    
    /**
     * Determine the final decision based on risk score
     */
    private String determineDecision(double riskScore) {
        if (riskScore < 0.3) {
            return "APPROVE";
        } else if (riskScore < 0.7) {
            return "REVIEW";
        } else {
            return "DECLINE";
        }
    }
    
    /**
     * Generate recommendations based on analysis
     */
    private void generateRecommendations(double riskScore, RiskLevel riskLevel, List<String> recommendations) {
        switch (riskLevel) {
            case LOW:
                recommendations.add("Process transaction normally");
                break;
            case MEDIUM:
                recommendations.add("Additional verification recommended");
                recommendations.add("Monitor account for suspicious activity");
                break;
            case HIGH:
                recommendations.add("Manual review required");
                recommendations.add("Contact customer for verification");
                recommendations.add("Consider temporary account restrictions");
                break;
        }
    }
    
    // Helper methods (simplified implementations for demo purposes)
    
    private void updateTransactionHistory(Transaction transaction) {
        accountTransactionHistory
            .computeIfAbsent(transaction.getAccountId(), k -> new ArrayList<>())
            .add(transaction);
    }
    
    private List<Transaction> getRecentTransactions(String accountId, int minutes) {
        List<Transaction> transactions = accountTransactionHistory.get(accountId);
        if (transactions == null) return new ArrayList<>();
        
        LocalDateTime cutoff = LocalDateTime.now().minus(minutes, ChronoUnit.MINUTES);
        return transactions.stream()
            .filter(t -> t.getTransactionTimestamp().isAfter(cutoff))
            .toList();
    }
    
    private boolean isHighRiskCountry(String country) {
        // Simplified implementation
        Set<String> highRiskCountries = Set.of("XX", "YY", "ZZ"); // Placeholder countries
        return highRiskCountries.contains(country);
    }
    
    private boolean hasGeographicInconsistency(Transaction transaction) {
        // Simplified implementation - check if location differs significantly from recent transactions
        return false; // Placeholder implementation
    }
    
    private boolean isNewDevice(Transaction transaction) {
        // Simplified implementation
        return true; // Placeholder - would check against known devices
    }
    
    private boolean isSuspiciousIP(String ipAddress) {
        // Simplified implementation
        return ipAddress != null && ipAddress.startsWith("10.0.0"); // Placeholder logic
    }
    
    private boolean isHighRiskMerchantCategory(String category) {
        // High-risk categories
        Set<String> highRiskCategories = Set.of("gambling", "adult", "cryptocurrency", "money_transfer");
        return highRiskCategories.contains(category);
    }
}
